beater/beater.go

```diff
@@ -618,12 +620,31 @@ func newTransformConfig(beatInfo beat.Info, cfg *config.Config) (*transform.Conf
 }
 
 func newSourcemapStore(beatInfo beat.Info, cfg config.SourceMapping) (*sourcemap.Store, error) {
-	esClient, err := elasticsearch.NewClient(cfg.ESConfig)
+	logger := logp.NewLogger(logs.Sourcemap)
+	if len(cfg.SourceMapConfigs) == 0 {
+		esClient, err := elasticsearch.NewClient(cfg.ESConfig)
+		if err != nil {
+			return nil, err
+		}
+		index := strings.ReplaceAll(cfg.IndexPattern, "%{[observer.version]}", beatInfo.Version)
+		b := sourcemap.NewESStore(esClient, index, logger)
+
+		return sourcemap.NewStore(b, logger, cfg.Cache.Expiration)
+	}
+
+	// make a copy of the default http client, configure a Transport to be
+	// the same as the esClient, instrument it, and set it on the http
+	// client.
+	c := *http.DefaultClient
+	tr, err := elasticsearch.NewHTTPTransport(cfg.ESConfig)
```

Comment:
Communicating with fleet-server may require a different TLS config (e.g. different CA certificate), so I don't think using the elasticsearch transport is the right thing to do. I don't know where we get that config from though, we'll need to investigate and/or ask for advice.

Reply:
Asked folks working on fleet, waiting for a confirmation. Initial response is it might be handed to the child proc during the gRPC handshake?

Reply:
OK so after some digging, I found that elastic-agent is sending the fleet config to apm-server along with the apm-server specific bits. Initially I thought it wasn't because it doesn't show up in `elastic-agent inspect`.

Example:

```json
{
  "apm-server": {
    "api_key": {
      "enabled": null,
      "limit": null
    },
    "capture_personal_data": null,
    "default_service_environment": null,
    "host": ":8200",
    "max_event_size": null,
    "rum": {
      "allow_headers": null,
      "allow_origins": null,
      "allow_service_names": null,
      "enabled": null,
      "event_rate": {
        "limit": null,
        "lru_size": null
      },
      "exclude_from_grouping": null,
      "library_pattern": null,
      "response_headers": null,
      "source_mapping": {
        "elasticsearch": {
          "api_key": null
        }
      }
    },
    "secret_token": null
  },
  "data_stream": {
    "namespace": "default"
  },
  "fleet": {
    "access_api_key": "OE5WWkEzb0JhR19YNFNhdUJuajg6eWJIWEdQV25TMGl6VEF0Wjk4YTlLQQ==",
    "agent": {
      "id": "ab5f009b-1497-46ce-b799-6fa505327b41",
      "logging": {
        "level": "info"
      },
      "monitoring": {
        "http": {
          "enabled": false,
          "host": "",
          "port": 6791
        }
      }
    },
    "enabled": true,
    "host": {
      "id": "d32ae199d8deb2d5edeae9b6a8434d89"
    },
    "hosts": [
      "172.20.0.4:8220"
    ],
    "protocol": "https",
    "reporting": {
      "check_frequency_sec": 30,
      "threshold": 10000
    },
    "ssl": {
      "certificate_authorities": [
        "/etc/pki/tls/certs/fleet-ca.pem"
      ],
      "renegotiation": "never",
      "verification_mode": "full"
    },
    "timeout": "10m0s"
  },
  "id": "3d3b4503-5c02-4e12-9715-5acd264205c1",
  "meta": {
    "package": {
      "name": "apm",
      "version": "0.3.0"
    }
  },
  "name": "apm",
  "revision": 1,
  "type": "apm",
  "use_output": "default"
}
```

We'll need to update `beater/config.Fleet` to add the fleet bits we need (`hosts`, `protocol`, `ssl`, `access_api_key`), and probably remove `kibana` while we're there: https://github.com/elastic/apm-server/blob/7f0fafec377e6ac39aa758f681cab785e4d5eeea/beater/config/integration.go#L68-L70